feat: PWA pull to refresh

Soxasora opened the pull request:
## Description

Pull-to-refresh for iOS, completing #1258.

As `router.reload()` is not graceful and `router.replace(...)` adds "/~" to the path breaking functionality across the app, I've decided to use `router.push()` instead, reloading the same page.
This updates the browser's history but since we're talking about a PWA there's no history to be accessed from the user.

This component wraps `<Layout />` and is present in every page that make use of it.

## Screenshots

https://github.com/user-attachments/assets/46209199-ae9d-433b-8c5f-92ae9f073c6c

## Additional Context

At the moment it's enabled also for Android, disabling its native pull-to-refresh. If we want to enable it back we can pass `android=false`

Not being a master of creativity and CSS I want to know if you'd like to see something else instead of _refreshing..._, even though I think it's geeky in its own way

## Checklist

**Are your changes backwards compatible? Please answer below:**
checkPWA prevents it from being triggered on anything that's not a PWA, also being a new component that has minimal CSS and loads only if the user is at the top of the page, I can say that it doesn't break functionality.

**On a scale of 1-10 how well and how have you QA'd this change and any features it might affect? Please answer below:**
7, I don't have an Android phone and I would like to know if there are any problems with it, as said before we can disable or enable Android's PTR.
As it should work only in a PWA context, I also tested it on browsers

**For frontend changes: Tested on mobile? Please answer below:**
Yes, being a mobile-only feature

**Did you introduce any new environment variables? If so, call them out explicitly here:** No

ekzyis commented on components/pull-to-refresh.js:
```diff
@@ -0,0 +1,95 @@
+  }, [isPWA, isAndroid, android])
+
+  const handleTouchMove = useCallback((e) => {
+    if (touchStartY.current === 0) return
```
Is this needed so the text doesn't get stuck in some cases? At least that's what I can tell starts happening when I remove this line. A comment in the code would be nice here.

Soxasora replied:
Yes! I thought I was going crazy because between refreshes the text would get stuck, noted thanks :P

ekzyis commented on components/pull-to-refresh.js:
```diff
@@ -0,0 +1,95 @@
+  }, [])
+
+  const handleTouchEnd = useCallback(() => {
+    if (touchStartY.current === 0 || touchEndY.current === 0) return
```
Same reason here I suppose

Soxasora replied:
Touch can be finicky with JS, I wanted to make sure that we can't encounter weird behavior

ekzyis commented on components/pull-to-refresh.js:
```diff
@@ -0,0 +1,95 @@
+import { useState, useRef, useEffect, useCallback } from 'react'
+import styles from './pull-to-refresh.module.css'
+
+const PULL_THRESHOLD = 300
```
100 as twice the threshold when `pull to refresh` starts showing might be a better threshold.

I would also call this `REFRESH_THRESHOLD` since `PULL_THRESHOLD` sounds like that's when we show the first message.

Soxasora replied:
Noted, you're right it would make much more sense!

ekzyis commented on components/pull-to-refresh.js:
```diff
@@ -0,0 +1,95 @@
+        {pullDistance > 0 || isRefreshing
+          ? (
+            <>
+              <p className={`${styles.pullMessage} ${pullDistance > 50 || isRefreshing ? styles.fadeIn : ''}`}>
```
This suggests you want to show the message when `pullDistance > 50`, but above you're checking for `pullDistance > 0`. Isn't checking it at one place enough?

Soxasora replied:
Ok so, I think that stayed there during multiple versions of this code, IIRC before placing touch checks this was a workaround to get me through refreshes and avoid unintended messages

ekzyis commented on components/pull-to-refresh.js:
```diff
@@ -0,0 +1,95 @@
+              <p className={`${styles.pullMessage} ${pullDistance > 50 || isRefreshing ? styles.fadeIn : ''}`}>
+                {getPullMessage()}
+              </p>
+              {isRefreshing && <div className={styles.spacer} />}
```
I think we should also drop this since afaict, it only causes layout shift

Soxasora replied:
And it doesn't look clean

ekzyis commented on components/pull-to-refresh.module.css:
```diff
@@ -0,0 +1,25 @@
+    top: 0;
+}
+
+.spacer {
+    position: relative;
+    margin-bottom: 32px;
+}
```
see other comment

ekzyis commented on components/pull-to-refresh.module.css:
```diff
@@ -0,0 +1,25 @@
+    left: 50%;
+    top: -50px;
+    transform: translateX(-50%);
+    font-size: 18px;
```
```suggestion
    font-size: small;
```

ekzyis reviewed (approved):
Hey, this is very impressive! It works great and the code is very clean. Stackers (myself included) are going to love this! Additionally, this was a breeze to review. Thank you so much for this! Looking forward to more contributions from you :eyes: 

Now to the review:

I tested this on desktop by simply manually setting `isPWA` to `true` and then using the mobile emulator. I also tested that the refresh actually works by loading new comments that I created in a new tab.

I only have a few UX and CSS nitpicks:

1. is it possible to prevent scrolling the window down when the `pull to refresh` shows up and I want to abort?

https://github.com/user-attachments/assets/378a4f32-06e6-468a-af34-7dbc75ba045e

2. Why did you add the spacer during refresh? Afaict it does nothing except create layout shift. Therefore I think we should also remove it, it's even more cleaner that way. So the meme should have actually been this:

![image](https://github.com/user-attachments/assets/5a54265b-d2d8-44b1-a0a2-8abe6646215a)

3. the pull message is a little big so I think we should use `font-size: small` in the `pullMessage` CSS class.

4. the distance to pull down might be a little too much (almost half of the screen) but we can roll this out and then see if we need to adjust it.

5. pulling down also works when I start to pull down from the bottom of the screen when I am scrolled all the way to the top. Not sure if that's how other apps do it, I think you need to start at the top?

After dropping `refreshing ...`, the spacer and making the text smaller, the UI/UX is this:

https://github.com/user-attachments/assets/5d7ad248-834b-4467-a33f-f82c748c6059

Regarding Android: I haven't tested it on Android but since this works so well, I think we can keep your code enabled on Android to have a consistent experience.

If anything is weird, it's very easy to go back to the native behavior on Android.

Since these are all just nitpicks and it's already as-is a huge improvement, this is approved from my side :white_check_mark: 

<sub>@huumn is the final boss though</sub>

Soxasora commented on components/pull-to-refresh.js:
```diff
@@ -0,0 +1,95 @@
+    setIsPWA(androidPWA || iosPWA)
+  }
+
+  useEffect(() => {
+    checkPWA()
+  }, [])
```
Oops 😳

Soxasora commented:
> > is it possible to prevent scrolling the window down when the pull to refresh shows up and I want to abort?
> 
> a solution for this might be to use [the behavior in your first video](https://github.com/stackernews/stacker.news/issues/1258#issuecomment-2480071902) where the whole page moves down when you pull down

The behavior you're encountering should be uniquely present on desktop as it doesn't allow overscroll like mobile devices, if you notice in my two videos, the whole page moves down when you pull down :P

wait- now that I think about it, to make it theoretically work on Android I conditionally disable overscroll, we would get the same result you get on desktop... we definitely need to go with your UX/UI changes in order to make it pretty for both devices

Soxasora commented:
https://github.com/user-attachments/assets/2fc2a22d-fd46-4a02-b7c8-f52c58b388fc

well it looks super cool.

I moved `overscroll-behavior-y` into the body to avoid double refreshes, applied changes and corrected some typos

Soxasora commented:
i forgot to memoize the messages 🫣
with calculated margin we can stay safe across devices, didn't cross my mind, thanks @huumn!
